rclcpp/test/rclcpp/test_executor.cpp

```diff
@@ -294,6 +294,41 @@ TEST_F(TestExecutor, spin_some_elapsed) {
   ASSERT_TRUE(timer_called);
 }
 
+TEST_F(TestExecutor, spin_for_duration)
+{
+  DummyExecutor dummy;
+  auto node = std::make_shared<rclcpp::Node>("node", "ns");
+  bool timer_called = false;
+  auto timer =
+    node->create_wall_timer(
+    std::chrono::milliseconds(10), [&]() {
+      std::this_thread::sleep_for(std::chrono::milliseconds(10));
+      timer_called = true;
+    });
+  dummy.add_node(node);
+  // Wait for the wall timer to have expired.
+  std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  dummy.spin_for(std::chrono::seconds(1));
+
+  ASSERT_TRUE(timer_called);
+}
```

Comment:
This kind of timing test is actually very fragile since we don't use real-time operating systems on our buildfarm machines.

I would not be surprised to see this become a flakey test at some point.

---

To help avoid that, consider:

- using `0` for the timer's duration (rather than 10ms)
- do not sleep in the callback for the timer (not sure what the purpose of that is)
- do not sleep to wait for the timer to expire, check `timer->time_until_trigger() <= 0` in a busy-loop (can sleep inside the busy loop)

---

Taking a step back, what is being tested here? That `spin_for` calls callbacks that were ready when it is entered? If so, maybe use a duration of `0` for `sleep_for()`? As this should work even for non-blocking spins.